src/database-controller/src/core/framework.js

```diff
@@ -0,0 +1,421 @@
+        .catch(ignoreError);
+    this._dockerSecretDef &&
+      k8s
+        .patchSecretOwnerToFramework(this._dockerSecretDef, frameworkResponse)
+        .catch(ignoreError);
+  }
+
+  silentDelete() {
+    // do not await for delete
+    this._configSecretDef &&
+      k8s.deleteSecret(this._configSecretDef.metadata.name).catch(ignoreError);
+    this._priorityClassDef &&
+      k8s
+        .deletePriorityClass(this._priorityClassDef.metadata.name)
+        .catch(ignoreError);
+    this._dockerSecretDef &&
+      k8s.deleteSecret(this._dockerSecretDef.metadata.name).catch(ignoreError);
+  }
+
+  getUpdate() {
+    const update = {};
+    if (this._configSecretDef) {
+      update.configSecretDef = JSON.stringify(this._configSecretDef);
+    }
+    if (this._priorityClassDef) {
+      update.priorityClassDef = JSON.stringify(this._priorityClassDef);
+    }
+    if (this._dockerSecretDef) {
+      update.dockerSecretDef = JSON.stringify(this._dockerSecretDef);
+    }
+    return update;
+  }
+}
+
+async function synchronizeCreate(snapshot, addOns) {
+  await addOns.create();
+  try {
+    const response = await k8s.createFramework(snapshot.getRequest(false));
+    // framework is created successfully.
+    const frameworkResponse = response.body;
+    // don't wait for patching
+    addOns.silentPatch(frameworkResponse);
+    return frameworkResponse;
+  } catch (err) {
+    if (err.response && err.response.statusCode === 409) {
+      // doesn't delete add-ons if 409 error
+      logger.warn(`Framework ${snapshot.getName()} already exists.`);
+      throw err;
+    } else {
+      // delete add-ons if 409 error
+      addOns.silentDelete();
+      throw err;
+    }
+  }
+}
+
+async function synchronizeModify(snapshot) {
+  const response = await k8s.patchFramework(
+    snapshot.getName(),
+    snapshot.getRequest(false),
+  );
+  const frameworkResponse = response.body;
+  return frameworkResponse;
+}
+
+async function synchronizeRequest(snapshot, addOns) {
+  // any error will be raised
+  // if succeed, return framework from api server
+  // There may be multiple calls of synchronizeRequest.
+  // Poller and write-merger uses this method.
+  try {
+    await k8s.getFramework(snapshot.getName());
+    // if framework exists
+    const frameworkResponse = await synchronizeModify(snapshot);
+    logger.info(
+      `Request of framework ${snapshot.getName()} is successfully patched.`,
+    );
+    return frameworkResponse;
+  } catch (err) {
+    if (err.response && err.response.statusCode === 404) {
+      const frameworkResponse = await synchronizeCreate(snapshot, addOns);
+      logger.info(
+        `Request of framework ${snapshot.getName()} is successfully created.`,
+      );
+      return frameworkResponse;
+    } else {
+      throw err;
+    }
+  }
+}
+
+function silentSynchronizeRequest(snapshot, addOns) {
+  // any error will be ignored
+  synchronizeRequest(snapshot, addOns).catch(ignoreError);
+}
```

Comment:
At least log error

Reply:
Actually `ignoreError` does the logging thing.  I change the function name `ignoreError` to `logError` to make it more clear.